src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -0,0 +1,955 @@
+            }
+
+            registration.handle(
+                    packet.payload(),
+                    new PlayPayloadContext(
+                            playPacketListener::send,
+                            new ClientPacketHandler(playPacketListener),
+                            new EventLoopSynchronizedWorkHandler<>(playPacketListener.getMainThreadEventLoop(), packet.payload()),
+                            PacketFlow.CLIENTBOUND,
+                            listener.getConnection().channel().pipeline().lastContext(),
+                            Optional.empty()));
+        } else {
+            LOGGER.error("Received a modded custom payload packet from a server that is not in the configuration or play phase. Disconnecting server.");
+            throw new IllegalStateException("A server send a packet while not in the configuration or play phase. Somebody changed the phases known to NeoForge!");
+        }
+    }
+
+    /**
+     * Invoked by the server when it completes the negotiation with the client during the configuration phase.
+     * <p>
+     * This method determines what the versions of each of the channels are, and checks if the client and server have a compatible set of network channels.
+     * </p>
+     * <p>
+     * If the negotiation fails, a custom packet is send to the client to inform it of the failure, and which will allow the client to disconnect gracefully with an indicative error screen.
+     * </p>
+     * <p>
+     * This method should only be invoked for modded connections. Use {@link #onVanillaConnectionDetectedAtServer(ServerConfigurationPacketListener)} to indicate that during the configuration phase of the network handshake between a client and the server, a vanilla connection was detected.
+     * </p>
+     *
+     * @param sender        The listener which completed the negotiation.
+     * @param configuration The configuration channels that the client has available.
+     * @param play          The play channels that the client has available.
+     */
+    public void onModdedConnectionDetectedAtServer(ServerConfigurationPacketListener sender, Set<ModdedNetworkQueryComponent> configuration, Set<ModdedNetworkQueryComponent> play) {
+        final NegotiationResult configurationNegotiationResult = NetworkComponentNegotiator.negotiate(
+                knownConfigurationRegistrations.entrySet().stream()
+                        .map(entry -> new NegotiableNetworkComponent(entry.getKey(), entry.getValue().version(), entry.getValue().flow(), entry.getValue().optional()))
+                        .toList(),
+                configuration.stream()
+                        .map(entry -> new NegotiableNetworkComponent(entry.id(), entry.version(), entry.flow(), entry.optional()))
+                        .toList());
```

Comment:
Similar to the code deduplication in my other comment can we deduplicate the code between this and the play negotiation result? As the main thing would be adding a helper to call `NetworkComponentNegotiator#negotiate` with knownRegistrations and then the set of modded components

Reply:
Same as above, I tried merging them, but the code really got ugly and hard to read.

The duplication is annoying but not worse then the merged variant so I will leave it.